src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEFileSaveWatcher.cs

```diff
@@ -0,0 +1,45 @@
+﻿using System;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem.VS.Automation;
+using VSLangProj;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.TempPE
+{
+    /// <summary>
+    /// This class handles being notified whenever a source file is saved by the hierarchy
+    /// </summary>
+    [Export(typeof(IFileActionHandler))]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    internal class TempPEFileSaveWatcher : IFileActionHandler
+    {
+        private readonly VSBuildManager _buildManager;
+        private readonly Lazy<ITempPEBuildManager> _tempPEBuildManager;
+
+        [ImportingConstructor]
+        public TempPEFileSaveWatcher(
+            [Import(typeof(BuildManager))]VSBuildManager buildManager,
+            Lazy<ITempPEBuildManager> tempPEBuildManager)
+        {
+            _buildManager = buildManager;
+            _tempPEBuildManager = tempPEBuildManager;
+        }
+
+        public async Task<bool> TryHandleFileSavedAsync(IProjectTree tree, string newFilePath, bool saveAs)
+        {
+            // We don't want to be the one to realise the Lazy<ITempPEManager> so ignore changes in projects that don't need TempPE
+            if (!_buildManager.ProjectNeedsTempPE) return false;
```

Comment:
This feels like the wrong dependency; this class only a dependency on BuildManager just to ask if ITempPEBuildManager is actually doing work.

Reply:
This is related to https://github.com/dotnet/project-system/pull/4349#discussion_r239280321 and will be removed with that change.